// axios/lib/dist/axios.dev.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _utils = _interopRequireDefault(require("./utils.js"));

var _bind = _interopRequireDefault(require("./helpers/bind.js"));

var _Axios = _interopRequireDefault(require("./core/Axios.js"));

var _mergeConfig = _interopRequireDefault(require("./core/mergeConfig.js"));

var _index = _interopRequireDefault(require("./defaults/index.js"));

var _formDataToJSON = _interopRequireDefault(require("./helpers/formDataToJSON.js"));

var _CanceledError = _interopRequireDefault(require("./cancel/CanceledError.js"));

var _CancelToken = _interopRequireDefault(require("./cancel/CancelToken.js"));

var _isCancel = _interopRequireDefault(require("./cancel/isCancel.js"));

var _data = require("./env/data.js");

var _toFormData = _interopRequireDefault(require("./helpers/toFormData.js"));

var _AxiosError = _interopRequireDefault(require("./core/AxiosError.js"));

var _spread = _interopRequireDefault(require("./helpers/spread.js"));

var _isAxiosError = _interopRequireDefault(require("./helpers/isAxiosError.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Create an instance of Axios
 *
 * @param {Object} defaultConfig The default config for the instance
 *
 * @returns {Axios} A new instance of Axios
 */
function createInstance(defaultConfig) {
  var context = new _Axios["default"](defaultConfig);
  var instance = (0, _bind["default"])(_Axios["default"].prototype.request, context); // Copy axios.prototype to instance

  _utils["default"].extend(instance, _Axios["default"].prototype, context, {
    allOwnKeys: true
  }); // Copy context to instance


  _utils["default"].extend(instance, context, null, {
    allOwnKeys: true
  }); // Factory for creating new instances


  instance.create = function create(instanceConfig) {
    return createInstance((0, _mergeConfig["default"])(defaultConfig, instanceConfig));
  };

  return instance;
} // Create the default instance to be exported


var axios = createInstance(_index["default"]); // Expose Axios class to allow class inheritance

axios.Axios = _Axios["default"]; // Expose Cancel & CancelToken

axios.CanceledError = _CanceledError["default"];
axios.CancelToken = _CancelToken["default"];
axios.isCancel = _isCancel["default"];
axios.VERSION = _data.VERSION;
axios.toFormData = _toFormData["default"]; // Expose AxiosError class

axios.AxiosError = _AxiosError["default"]; // alias for CanceledError for backward compatibility

axios.Cancel = axios.CanceledError; // Expose all/spread

axios.all = function all(promises) {
  return Promise.all(promises);
};

axios.spread = _spread["default"]; // Expose isAxiosError

axios.isAxiosError = _isAxiosError["default"];

axios.formToJSON = function (thing) {
  return (0, _formDataToJSON["default"])(_utils["default"].isHTMLForm(thing) ? new FormData(thing) : thing);
};

var _default = axios;
exports["default"] = _default;
