// axios/dist/index.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toFormData = exports.spread = exports.isAxiosError = exports.Cancel = exports.all = exports.VERSION = exports.CancelToken = exports.isCancel = exports.CanceledError = exports.AxiosError = exports.Axios = exports["default"] = void 0;

var _axios = _interopRequireDefault(require("./lib/axios.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// Keep top-level export same with static properties
// so that it can keep same with es module or cjs
var Axios = _axios["default"].Axios,
    AxiosError = _axios["default"].AxiosError,
    CanceledError = _axios["default"].CanceledError,
    isCancel = _axios["default"].isCancel,
    CancelToken = _axios["default"].CancelToken,
    VERSION = _axios["default"].VERSION,
    all = _axios["default"].all,
    Cancel = _axios["default"].Cancel,
    isAxiosError = _axios["default"].isAxiosError,
    spread = _axios["default"].spread,
    toFormData = _axios["default"].toFormData;
exports.toFormData = toFormData;
exports.spread = spread;
exports.isAxiosError = isAxiosError;
exports.Cancel = Cancel;
exports.all = all;
exports.VERSION = VERSION;
exports.CancelToken = CancelToken;
exports.isCancel = isCancel;
exports.CanceledError = CanceledError;
exports.AxiosError = AxiosError;
exports.Axios = Axios;
var _default = _axios["default"];
exports["default"] = _default;
